Added message format checker

MrPointer opened the pull request:
This work adds support for a **Format Checker**, which accepts a message represented by a C-String and checks whether it fits the expected format of an NMEA message.

This work also comes with a handful of unit tests which should provide full coverage.
It's also well-documented.

@RonAmihai One thing I'd like to get your attention on is the way the tests are written.
I've left a `todo` note where a factory may be needed, but I couldn't decide on my own.

MrPointer committed:
Configured basic cmake project struct

MrPointer committed:
Configured basic catch-based test env

MrPointer committed:
Created first test and a basic implementation of a format-checker.

MrPointer committed:
Refactored a bit, added header to define protocol format.

MrPointer committed:
Added delimiter position tests, implemented appropriate checks

MrPointer committed:
Added protocol ending check (stop chars), tested as well.

MrPointer committed:
Added support for Catch as a system package
There's "backward-compatibility" in case the package can't be found.

MrPointer committed:
Covered null-string case in format-checker.

MrPointer committed:
Added coverage for message length

MrPointer committed:
Added message checksum validation
Also added some utility modules.

MrPointer committed:
Modified format-checker to cover invalid message types
This is instead of the previous coverage for explicit misplaced protocol delimiters of various types.

MrPointer committed:
Added coverage for data-less messages
Also refactored and renamed some protocol-format-related macro definitions.

MrPointer committed:
Explicitly defined compiler IDs on lib target
Also removed unnecessary inner `project` definitions.

MrPointer committed:
Fixed merge-issues and applied some refactoring
Also fixed false usage of `strcpy_s` on MSVC compilers.

MrPointer committed:
Refactored `FormatChecker` entirely
Mostly separated into smaller helper functions.

MrPointer committed:
Applied some basic refactoring on 'formatChecker' tests
Also applied tiny patches on the actual source.

MrPointer committed:
Added coverage for invalid checksum data (Not hex)

MrPointer committed:
Modified message-type checks, added tests for invalid lengths

MrPointer committed:
Added coverage for Talker-IDs

MrPointer commented:
I've added missing coverage, which also caused to modify some checks to be even "smarter".

RonAmihai commented:
Feedback about the code itself (without tests):

1. EVALID code should be returned instead of explicit 0
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L59

2. Unnecessary 'else-if' (first 'if' includes 'return' statement & 'else' doesn't improves readability anyways) 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L87

3. Consider to name out parameters with 'out' prefix to improve readability 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L93 

4. Use contants (defines) for offsets & length instead of explicit values
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L37
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L155
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L107
  
5. Use conteants (defines) for shifting values. They can be understood eventually but it is desirable to specify their vocation explicitly for readability 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/utility/GlobalUtils.c#L39

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+    if (errorCode)
+        return errorCode;
+
+    return 0;
```
EVALID code should be returned instead of explicit 0

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+        else
+            return -EMISSING_MESSAGE_DATA;
+    }
+    else if (messageSize > MESSAGE_MAX_LENGTH)
```
Unnecessary 'else-if' (first 'if' includes 'return' statement & 'else' doesn't improves readability anyways)

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+    return EVALID;
+}
+
+static int8_t validateChecksumFormat(const char *message, size_t messageSize, size_t *checksumIdealIndex)
```
Consider naming out-parameters with 'out' prefix to improve readability

MrPointer replied:
I disagree. The **out** modifier is mostly known to .NET developers (C# specifically), yet it's not very popular in C.
With that being said, I do agree that this function isn't very consistent with its' name, and that should be addressed - Either by separating code somehow or by renaming it.

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+        return errorCode;
+
+    // Validate message ending fits protocol
+    if (rawMessage[messageSize - 2] != PROTOCOL_STOP_CHAR_1 ||
```
Use constants (defines) for offsets & length instead of explicit values

MrPointer replied:
In this case we're merely checking the 2 last elements of the array, I don't think there's a need to name it somehow - Care to tell me if I'm wrong?

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+            return -EMISSING_CHECKSUM;
+    }
+
+    if (!isHex(message[*checksumIdealIndex + 1]) || !isHex(message[*checksumIdealIndex + 2]))
```
Use constants (defines) for offsets & length instead of explicit values

MrPointer replied:
Took a better approach by iterating over the checksum's length - That way it'll work even in extreme cases

RonAmihai commented on src/format/FormatChecker.c:
```diff
@@ -0,0 +1,167 @@
+    char receivedChecksum[3] = {0};
+
+    // ToDo: Optimize?
+    memcpy(receivedChecksum, message + checksumDelimiterIndex + 1, 2);
```
Use constants (defines) for offsets & length instead of explicit values

RonAmihai commented on src/utility/GlobalUtils.c:
```diff
@@ -0,0 +1,55 @@
+    int ret = 0;
+    while (*hexString && ret >= 0)
+    {
+        ret = (ret << 4) | hextable[*hexString++];
```
Use constants (defines) for shifting values. They can be understood eventually but it is desirable to specify their vocation explicitly for readability

RonAmihai reviewed (commented):
Review of the code itself (without tests)

RonAmihai commented:
As for the test - only two things to mention: 
1) Consider using constants (defines) also for string-append indexers. 
2) Factories would be nice addition whenever you mentioned.  

Other than that - the tests are written good overall.

MrPointer committed:
Converted message beginning & ending check to macro

MrPointer committed:
Modified checksum delimiter find process in format check

MrPointer committed:
Moved hex functions to separate header

MrPointer committed:
Separated checksum validation function to 'Validator' module

MrPointer committed:
Added 'FFF' mocking framework

Provided solution for C-mocking on-demand, allowing mocking a function only when required, while using the original otherwise, all inside a single executable.
This can be seen by the `ChecksumValidatorTests` example.

MrPointer committed:
Converted valid-message test to mock-based unit test

Original test will become an integration test eventually.

MrPointer committed:
Separated checksum validator to message-specific and general validators

MrPointer committed:
Reverted checksum module separation

Functions are still separated however, and are part of the same module.

MrPointer committed:
Separated checksum format validation to another module

MrPointer committed:
Separated checksum value comparison to another module

MrPointer committed:
Reflected latest changes in tests

MrPointer committed:
Added vscode launch config file

MrPointer committed:
Added fakes for newly separated checksum modules

MrPointer committed:
Added mocked tests for message checksum validator

MrPointer committed:
Added checksum format validation tests

Modified the function itself, optimizing for performance and avoiding unnecessary overhead checks.

MrPointer committed:
Moved checksum calc and comp modules to '/etc/checksum'

Also renamed 'ChecksumUtils' to 'ChecksumCalculator', as currently its the only thing it does + It better separates concerns and turns it into a valid standalone module.

MrPointer committed:
Added checksum calculation tests

Fixed a few iteration bugs in the checksum calculator.

MrPointer committed:
Added invalid input tests, fixed null-check bug

MrPointer committed:
Added checksum calculation 'overloads'

These are functions with similar purpose, only accepting different parameters. Both ar made inline (defined in header) and simply forward the call to original function.

MrPointer committed:
Modularized checksum-calc function to use 'StringUtils'

MrPointer committed:
Separated test scenarios for each overload

MrPointer committed:
Added remaining checksum overloads tests

MrPointer committed:
Added mock support for checksum calculation functions

MrPointer committed:
Added tests for 'checksumValueComparator'

MrPointer committed:
Renamed 'START' & 'STOP' to 'HEADER' & 'FOOTER'

MrPointer committed:
Moved test sources under 'src' and 'include' dirs

MrPointer committed:
Moved all 'DISABLE_FAKE' calls to top of scenario

MrPointer committed:
Made current impl of 'HexUtils' Performance-Optimized

It may affect space consumption a bit, so a different space-optimized version would be provided later.
Also added tests for the module.

MrPointer committed:
Added substring-to-hex utility function

Modified `ChecksumValueComparator` to use this instead

MrPointer committed:
Reorganized tests
